cmake: linker: make passing -no-pie configurable

dcpleung opened the pull request:
(See individual commits for details.)

The main issue is to properly pass `-no-pie` (or `--no-pie`) to the linker. Old binutils' ld (<= 2.36, used in Ubuntu 20.04) and XCC's xt-ld both do not support '-no-pie' in linker. If you pass '-no-pie' to them, they will be parsed separately as `-n` and `-o-pie`, which results in output file named `-pie` instead of the usual `zephyr*.elf` (and thus failing intermediate build steps). The rest of the commits are to make llvm/clang works in CI and to fix a partially non-functional `CONFIG_LLVM_USE_LD`.

andyross reviewed (dismissed):
This definitely gets my vote

aborisovich reviewed (dismissed):
And my axe 😆

tejlmand commented on cmake/linker/ld/target_base.cmake:
```diff
@@ -11,10 +11,15 @@ macro(toolchain_ld_base)
     ${TOOLCHAIN_LD_FLAGS}
   )
 
+  if(CONFIG_BOARD_NATIVE_POSIX)
```
first comment.
We should not filter this based on a board type.
It's to limiting regarding users ability to enable / disable this flag.

As minimum this should have a Kconfig and with a proper default value.
A given board can then define a different default.

The setting may be promptless to avoid users accidentially changing the setting.

tejlmand commented on cmake/linker/ld/target_base.cmake:
```diff
@@ -11,10 +11,15 @@ macro(toolchain_ld_base)
 
+  if(CONFIG_BOARD_NATIVE_POSIX)
+    zephyr_ld_options(
+      ${LINKERFLAGPREFIX},-no-pie
```
your commit message states:
> Zephyr SDK and others are, currently, not to produce PIE binaries by default.

I get that if Zephyr SDK and other currently do not produce PIE binaries per default, but what is the problem in specifying the flag in those cases then ?

Also, we are explicitly requesting the compiler to not do position independent code for `C` code here:
https://github.com/zephyrproject-rtos/zephyr/blob/ab764aada99a420ef7fe75ec6ea8984391fae2a5/CMakeLists.txt#L376-L377

so having `-no-pie` for the linker seems logical to me in this case, even though the toolchain you mention does so per default.

Is your intention to be able to actually use `-pie` on the linker in some use-cases ?

Or should we actually change the handling of `pie` / `-no-pie`, so that it compiler and linker flags are set as properties and then applied based on configuration.
Similar to how things are done for `coverage`, where the are both compiler and linker flags specified under the `coverage` property, which is then applied based on Kconfig.
Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/d598e26de42cf9fc8ba4e9d048db33e0a82004bf/arch/common/CMakeLists.txt#L72-L75

tejlmand replied:
> so having `-no-pie` for the linker seems logical to me in this case, even though the toolchain you mention does so per default.
> 
> Is your intention to be able to actually use `-pie` on the linker in some use-cases ?

Answering my own question.
I see the need for not passing the flag here: https://github.com/zephyrproject-rtos/zephyr/pull/54333#issue-1566858160

So atm it seems to be specific to the Xtensa toolchain, `xcc`.
Which means the best approach forward is to have a `no_position_independent` property also on the linker, and the Xtensa toolchain can then set this property flag to empty in `https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/linker/ld/xcc/linker_flags.cmake` (assuming the xcc uses the ld linker)

galak replied:
I believe we need this with LLVM as well.  So it might not be xcc specific.

tejlmand reviewed (changes requested):
Need a bit more description to fully understand the use-case.

Also, this should be having corresponding Kconfig setting, and not rely on board testing directly in CMake.

tejlmand commented:
> Need a bit more description to fully understand the use-case.
> 
> Also, this should be having corresponding Kconfig setting, and not rely on board testing directly in CMake.

The Kconfig part may not be required.
Seems to be toolchain related, and hence the toolchain properties should be defined properly.

So that's first point.
Of course if a better user-facing handling of when to apply the properties is desired, then those should go through Kconfig as described. But the immediate problem described in #54333 doesn't seem to need it.

andyross commented:
> I get that if Zephyr SDK and other currently do not produce PIE binaries per default, but what is the problem in specifying the flag in those cases then ?

Others can comment with more expertise, but the core issue is that this flag isn't supported compatibly across all binutils/gcc/clang/lld variants.  And we don't really want it anyway, it's not a feature we enable at code generation so we don't want the linker to have to care about it.

The sole exception (I think! @yperess caught this originally) is on Debian when using the distro-provided clang/lld to produce native_posix binaries.  There, the toolchain is hard-wired to generate pic/pie by default, so we need the flag to the linker to unbreak things.  But that requires per-toolchain hackery.

I think.

dcpleung commented:
@tejlmand I have updated the PR to use linker property. I tried this previously but was having issue with using `zephyr_ld_options()`. Using `zephyr_link_libraries()` got past that issue.

marc-hb commented:
> There [on Debian], the toolchain is hard-wired to generate pic/pie by default, so we need the flag to the linker to unbreak things. But that requires per-toolchain hackery.

Stating the obvious: this was done to increase security
https://wiki.debian.org/Hardening/PIEByDefaultTransition
https://wiki.ubuntu.com/SecurityTeam/PIE

A lot of background information on these pages, including this (not) fun bit:

> Note that some of these packages have already taken patches upstream from distros that were using patched versions of GCC that enabled PIE by default and added a GCC option -nopie to disable it. However, upstream GCC has settled on the -no-pie option as the means for disabling PIE when GCC is configured to use it by default.

Similar headaches:
- https://github.com/mesonbuild/meson/pull/6415

dcpleung commented:
I think something is wrong in cmake scripts... the error said it was using `/usr/bin/ld`, but yet the LLD linker cmake files were parsed instead.

tejlmand commented:
> I tried this previously but was having issue with using `zephyr_ld_options()`.

Yes, because that function was made to be able to check if a flag is valid, so it only handles simply flags and not generator expressions.

> Using `zephyr_link_libraries()` got past that issue.

yes, and this is the function to use now, because code is updated to use linker properties.

> I think something is wrong in cmake scripts... the error said it was using /usr/bin/ld, but yet the LLD linker cmake files were parsed instead. 

Sounds strange. Will see if I can re-produce.

dcpleung commented:
> > I think something is wrong in cmake scripts... the error said it was using /usr/bin/ld, but yet the LLD linker cmake files were parsed instead.
> 
> Sounds strange. Will see if I can re-produce.

Turns out if you only have LLVM/clang/lld 15 installed, the symlink `/usr/bin/ld.lld` would not be there as it comes from package `lld` (which pulls in `lld-14`). Therefore, using `find_program()` to find `ld.lld` would not work since it's no in PATH. CMake then defaults to using `/usr/bin/ld`. I think this is what is causing the build failure. I have added a commit to expand the search to `LLVM_TOOLCHAIN_HOME/bin` (which basically is `TOOLCHAIN_HOME`).

However, another issue is that `CONFIG_LLVM_USE_LD=y` does not have any effect. I am not totally sure why... my guess is that the cmake scripts under `cmake/toolchain` is parsed before kconfigs, and none of `CONFIG_*` exist.

dcpleung commented:
Errr... the new commit doesn't seem to work with the Clang/LLVM workflow.

dcpleung committed:
linker: llvm: provide a default config file for clang

Some distros may provide config files for clang to change its
default behavior. We need to override that, or else developers
may be using different defaults and we will have confusing
bug reports in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

marc-hb commented:
Can you please rename the brand new `xcc-clang/` directory to `xt-clang/`?

There are two toolchains variants: `xt-xcc` (gcc-based) and `xt-clang`, so naming anything `xcc-clang` does not seem to make sense.

See
https://thesofproject.github.io/latest/introduction/index.html#toolchain-faq
https://github.com/thesofproject/sof/pull/7027#discussion_r1095792670

galak commented:
> Can you please rename the brand new `xcc-clang/` directory to `xt-clang/`?
> 
> There are two toolchains variants: `xt-xcc` (gcc-based) and `xt-clang`, so naming anything `xcc-clang` does not seem to make sense.
> 
> See https://thesofproject.github.io/latest/introduction/index.html#toolchain-faq [thesofproject/sof#7027 (comment)](https://github.com/thesofproject/sof/pull/7027#discussion_r1095792670)

That should be a follow PR since xcc-clang already exists in the tree.

marc-hb commented:
> That should be a follow PR since xcc-clang already exists [ELSEWHERE] in the tree.

Missed the other places sorry.

But... why couldn't this be fixed _before_? PIE flavors are complex already, "mashup" toolchain names don't really help IMHO.

It's not like anyone uses `xt-clang` yet, is it? Then better fix this quickly before anyone does and complains.

galak commented:
> But... why couldn't this be fixed _before_? PIE flavors are complex already, "mashup" toolchain names don't really help IMHO.
> 
> It's not like anyone uses `xt-clang` yet, is it? Then better fix this quickly before anyone does and complains.

The rename can come before or after, just shouldn't get mixed up in this PR.  Its 2 completely unrelated changes.

marc-hb commented:
> Some linkers (such as the one in xcc-clang) do not support this flag and will behave erratically if set (such as parsing it as "-n" and "-o").

I don't think this commit message makes sense. As found by @andyross in https://github.com/zephyrproject-rtos/zephyr/pull/54331#issuecomment-1416016501, `-no-pie` has been "ping-ponging around quite a bit in the last few months". If this fix can't be described clearly then the "ping-ponging" will surely continue.

First there is no such thing as `xcc-clang`, see above. I get that we don't want to fix the directory names yet but that does not require the commit messages and comments in the source to keep using the wrong name and spread the confusion even further. English text can start using the accurate name already. "xcc-clang" is especially confusing considering this issue at hand affects `xt-clang` but not `xt-xcc`.

A more important issue with this commit message is that **the linker in xt-clang 2022 has not changed**. `xt-clang` 2022 still uses GNU ld just like in xt-xcc did. So this PR is not fixing a problem with "some linkers", at least not as far as `xt-clang` is concerned. If anything it's more about `clang`'s inability to pass a GNU ld linker option to GNU ld without using `-Wl` 

> Its 2 completely unrelated changes.

This PR is trying to fix some compatibility issue(s) between "xcc" and "clang" so I think carrying forward a super confusing and wrong "xcc-clang" name that concatenates the two is actually quite relevant.

dcpleung commented:
Ok. I think I got it mostly figured out. The CI failure is due to https://github.com/zephyrproject-rtos/zephyr/issues/54720 (which `qemu_arc_em` failed `tests/kernel/mem_protect/syscalls/`).

nashif commented:
@evgeniy-paltsev / @ruuddw  can you verify on metaware?

ruuddw commented:
> can you verify on metaware?

@nashif  Not sure what your question is, test the CI failure on nSIM or check if this PR breaks anything for MetaWare?
Anyway, I think for both the answer is: no problems.

ruuddw commented:
@nashif MetaWare doesn't have the '-no-pie' option, but no_postion_independent is already handled by the arcmwdt cmake files. I don't see any conflicts. Thanks for checking!

aborisovich commented:
Hi there,
we have some issues when building SOF project on Windows with xt-clang toolchain and this PR.
I've cherry picked your commits and run a build on my PR https://github.com/thesofproject/sof/pull/7027 .
I've got following error log:
[output.log](https://github.com/zephyrproject-rtos/zephyr/files/10767585/output.log)
Uploading also part of my environment setup:
[env.log](https://github.com/zephyrproject-rtos/zephyr/files/10767586/env.log)

Please note, that on Windows we use MSYS to provide gcc implementation for Windows and other linux tools.
This is required by both Zephyr SDK and SOF project that builds firmware tools using native linux toolchains.
So when I set all the environmetal variables to select xt-clang toolchain, C:\msys64\usr\bin\ld.bfd.exe linker is used (what is not desirable since xt-clang linker should be selected).

aborisovich commented:
Ok, it looks like we need to set TOOLCHAIN_HOME env variable from now on to avoid those issues...

aborisovich commented:
Any update on this? We have a strong dependency to this PR.

dcpleung commented:
> Any update on this? We have a strong dependency to this PR.

Have you tried #54836? That should get rid of the `-no-pie` warning.

galak commented:
> Have you tried #54836? That should get rid of the `-no-pie` warning.

Would like to see this cleanup merged as need it form arm-llvm as well ;).

aborisovich commented:
> > Any update on this? We have a strong dependency to this PR.
> 
> Have you tried #54836? That should get rid of the `-no-pie` warning.

Thanks for pointing me in right direction!

tejlmand commented on cmake/bintools/gnu/ld_version.cmake:
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: Apache-2.0
```
this file should be converted into a `Findld.cmake` and placed in this folder:
https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/modules

that will allow us to call:
```
find_package(ld)
```
instead of `include(${ZEPHYR_BASE}/cmake/bintools/gnu/ld_version.cmake)`.

That will also follow the common CMake practice of providing a `<package_name>_VERSION` setting, as well as `<package_name>_FOUND`.
One can even do a `find_package(ld REQUIRED)` to ensure `ld` is found, and fail if not.

To consider:
- is `ld` the best name, or should it be `Gnuld` /  `binutils_ld`, or similar ?
I'm open to good naming suggestions on this.

**Note**: making a `Findld.cmake` will also allow us to update this code:
https://github.com/zephyrproject-rtos/zephyr/blob/0c9ca9aefaa1e6a54dfc23ce4e74bcd3c58dd3f3/cmake/linker/ld/target.cmake#L4-L13
to simply use `find_package(ld REQUIRED)`

marc-hb replied:
> is ld the best name, or should it be Gnuld / binutils_ld, or similar ?

It could be called `ld.bfd`, that's how both Debian and Fedora call it. I think many more people understand "GNU ld" however. `ld.bfd` would not be ambiguous but many people would have to google it.

```
$ md5sum /usr/bin/ld{,.bfd}
9b6cfc52ea4aa58adf87c961c4ae5a9d  /usr/bin/ld
9b6cfc52ea4aa58adf87c961c4ae5a9d  /usr/bin/ld.bfd
```

GNU ld should really not be called just `ld` because that's too generic and ambiguous. For instance the native, Mach-O linker on macOS is also called `ld`.

lld - LLVM
ld.gold
ld / ld.bfd - GNU ld
ld - Mach-O linker on macOS

dcpleung replied:
Let me rework the PR to support `find_package()`.

tejlmand commented on Kconfig.zephyr:
```diff
@@ -167,12 +167,11 @@ config CMAKE_LINKER_GENERATOR
 
-config LLVM_USE_LD
-	bool "LLVM use ld linker"
+config LLVM_USE_LLD
```
commit message states:
> This also changes CONFIG_LLVM_USE_LD to CONFIG_LLVM_USE_LLD.
> Since they are of opposite usage, it cannot utilize the usual
> deprecation process since that would use "select" to enable
> both, which cannot not be done in this case.

not understood ?
There is no `select LLVM_USE_LD`, there is only a `default y` in existing setting.

This can easily be changed into a choice, like this:
```
choice LLVM_LINKER
      prompt "LLVM Linker"
      depends on "${ZEPHYR_TOOLCHAIN_VARIANT}" = "llvm"
      default LLVM_USE_LD

config LLVM_USE_LD
      bool "LLVM use ld linker"
      help
        <existing help text>

config LLVM_USE_LLD
      bool "LLVM use lld linker"
      help
        <new help text>

endchoice
```
which will keep `ld` the default linker, and still provide a possibility to select `lld`

**note**: I have a feeling this code should actually be placed here instead:
https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/toolchain/llvm/Kconfig

tejlmand commented on cmake/linker/ld/target.cmake:
```diff
@@ -121,7 +128,7 @@ function(toolchain_ld_link_elf)
   )
 
-  if(${CMAKE_LINKER} STREQUAL "${CROSS_COMPILE}ld.bfd")
+  if(${CMAKE_LINKER} MATCHES "${CROSS_COMPILE}ld.bfd$")
```
note: `CROSS_COMPILE` may be an absolute path, so if this change is done because you want to handle cases where `CROSS_COMPILE` is just a string and `CMAKE_LINKER` is an absolute path, then new surprises could be lurking in the dark.

Could you comment a bit on your observations / reasons for the change ?
Also consider adding a comment in the CMake code, just to on safe side in case other starts to modify this.

dcpleung replied:
`${CMAKE_LINKER}` always comes back as full path to the executable on my tests. So a straight `STREQUAL` would never match. Let me try using `get_filename_component()` to extract the file name for comparison.

tejlmand reviewed (changes requested):
some more comments

dcpleung commented on cmake/linker/ld/target.cmake:
```diff
@@ -1,15 +1,22 @@
+    set(LD_SEARCH_PATH PATHS ${TOOLCHAIN_HOME} NO_DEFAULT_PATH)
+  endif()
+
+  find_program(CMAKE_LINKER ${CROSS_COMPILE}ld.bfd ${LD_SEARCH_PATH})
```
It's because of the `execute_process()` above. If that does not produce a usable linker, we need to find it ourselves.

dcpleung commented on cmake/linker/ld/target_base.cmake:
```diff
@@ -33,7 +33,7 @@ macro(toolchain_ld_base)
   endif()
 
-  if (CONFIG_LLVM_USE_LD)
+  if ("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "llvm")
```
I didn't know there were `cmake/toolchain/*/Kconfig` since both gcc and llvm don't have that. Will do that instead.

dcpleung committed:
toolchain: move CONFIG_LLVM_USE_LD into cmake/toolchain/llvm

This moves CONFIG_LLVM_USE_LD into cmake/toolchain/llvm as this
is a toolchain kconfig. Also make it a choice to allow the use
of LLVM's lld as linker.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: use find_package() to find LLVM lld

This introduces a new cmake module FindLlvmLld.cmake to do
the work to discover LLVM lld linker.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: defer setting variable LINKER till target tools

The variable LINKER is dependent on CONFIG_LLVM_USE_LLD
or CONFIG_LLVM_USE_LD, and these kconfigs are not
available when toolchain/llvm/generic.cmake is parsed.
So setting LINKER needs to be deferred to target.cmake
where kconfigs are available.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: ld: see if compiler has a preferred linker

This asks the compiler if it has its own preference for ld.bfd.
This is useful for LLVM (when CONFIG_LLVM_USE_LD=y) so we know
which linker clang is using.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: lld: see if clang has a preferred linker

This asks the clang if it has its own preference for ld.lld.
This is to mirror what we are doing to find GNU ld, and to
make sure we are using the linker clang is using.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: llvm: defer setting target triple to target.cmake

Since kconfigs are not available when generic.cmake is parsed.
Setting the target triple for x86 needs to be deferred to
target.cmake as it needs to know whether CONFIG_64BIT is
enabled. This also moves the ARM triple to target.cmake as
triple is needed for target tools.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
cmake: linker: make passing -no-pie configurable

This adds a new linker property specifically for passing
"-no-pie" to linker. Older binutils' LD (<= 2.36) do not
support this flag and will behave erratically if set. It
would parse "-no-pie" separately as "-n" and "-o-pie",
which would result in the output file being "-pie"
instead of "zephyr*.elf". Moreover, LLVM lld does not
support -no-pie but --no-pie (note the extra hyphen).
By having no-pie as a linker property, we can pass
correct no-pie flag to these linkers (or none at all).

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Hm... compliance check complains about `CONFIG_LLVM_*` because they are not available unless LLVM is used. Not sure if I should add these to the kconfig exempt list...

dcpleung commented:
I have updated the PR to use `find_package()` for both `ld` and `lld`.

dcpleung committed:
linker: ld: lld: do not do RELRO when using llvm/clang

GNU ld and LLVM lld both complain under C++:
  error: section: init_array is not contiguous with other relro sections

So do not create RELRO program header.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: ld: include crt{begin,end}.o if LIBGCC_DIR is defined

Only include crtbegin.o and crtend.o when LIBGCC_DIR is defined.
Since LIBGCC_DIR is not defined when compiling for posix
architecture, crt{begin,end}.o cannot be referred to via
LIBGCC_DIR.

Also note that, when using llvm/clang, crt{begin,end}S.o are
automatically for native_posix which collide with symbols in
crt{begin,end}.o. So there is no point in making LIBGCC_DIR
available for native_posix under llvm/clang.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tejlmand commented on cmake/linker/ld/target.cmake:
```diff
@@ -7,7 +7,9 @@ set(CMAKE_LINKER ${GNULD})
 if(NOT "${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "host")
-  if(CONFIG_CPP_EXCEPTIONS)
+  if(CONFIG_CPP_EXCEPTIONS AND
+     (EXISTS "${LIBGCC_DIR}/crtbegin.o") AND
+     (EXISTS "${LIBGCC_DIR}/crtend.o"))
```
To me this looks as a very fragile test.

Could we do this another / cleaner way ?

dcpleung replied:
I could change it to test if `arch` is `posix` and `CONFIG_LLVM_USE_LD` is enabled. I have only see it for `native_posix` as llvm links `crt{being,end}S.o` automatically. Let me try that.

dcpleung replied:
I removed this commit as it is no longer an issue on my machine. The error I had was when I did not have the full list of (other) commits. Having the other commits together probably have fixed the original issue.

dcpleung replied:
Changed this to test if `LIBGCC_DIR` is defined. Since posix arch does not have it, it skips including `crtbegin.o` and `crtend.o`.

tejlmand reviewed (changes requested):
Thanks, nice improvement.

Some general comments, but I've found no major issues.

dcpleung commented on cmake/modules/FindGnuLd.cmake:
```diff
@@ -17,15 +17,25 @@
-if(NOT GNULD)
-  find_program(GNULD ${CROSS_COMPILE}ld ${LD_SEARCH_PATH})
+  if(DEFINED TOOLCHAIN_HOME)
+    # When Toolchain home is defined, then we are cross-compiling, so only look
```
It was a directy copy from `cmake/toolchain/ld/target.cmake`. I will update the comment.

marc-hb commented on cmake/modules/FindGnuLd.cmake:
```diff
@@ -0,0 +1,59 @@
+    # - "GNU ld (Zephyr SDK 0.15.2) 2.38"
+    # - "GNU ld (Gentoo 2.39 p5) 2.39.0"
+    string(REGEX MATCH
+           "GNU ld \\(.+\\) ([0-9]+[.][0-9]+[.]?[0-9]*).*"
```
Maybe a bit stricter?
```suggestion
           "GNU ld \\(.+\\) ([0-9]+[.][0-9]+[.]?[0-9]*)$"
```
What happens when there's no match? Did you locally break the regex and see how that's reported?

dcpleung replied:
The version string would simply be empty. Cmake will not print the version, and any version comparisons will be false.

dcpleung committed:
linker: use find_package() to find GNU ld

This introduces a new cmake module FindGnuLd.cmake to do
the work to discover GNU ld (of binutils).

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

marc-hb commented on cmake/toolchain/llvm/target.cmake:
```diff
@@ -1,3 +1,25 @@
+  set(LINKER ld)
+elseif(CONFIG_LLVM_USE_LLD)
+  set(LINKER lld)
+else()
+  message(FATAL_ERROR
+          "Need to enable either CONFIG_LLVM_USE_LD or CONFIG_LLVM_USE_LLD!")
```
> Is this really needed ? A choice will always have a value,...

Unless when editing the .config file manually or generating from fragments and some non-Zephyr script?

> so the only way that both CONFIG_LLVM_USE_LD and CONFIG_LLVM_USE_LLD could n is in case someone extended the choice with an extra config option.

Could someone really add an extra KConfig option without editing this code?

> Finally, fatal errors should be used with absolute caution when it comes to Kconfig settings because of #9573

I only skimmed 9573 sorry but the main message seems to be "don't generate invalid .config files that menuconfig" can't read again. It will of course always be possible to make invalid .config edits manually.

Can this FATAL_ERROR really cause invalid .config files to be generated?

In my experience CMake and Kconfig errors tend to be a nightmare to debug + no one wants to go anywhere near them. Very understandable when you look at major design mistakes like COMMAND_ERROR_IS_FATAL false by default, MYPROGRAM_NOT-FOUND and other crazy and super time-consuming CMake stuff like https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/

So in general we need MORE user-friendly error messages, not fewer but maybe this particular case is different? Not convinced of that yet.

dcpleung replied:
If someone adds a new linker kconfig, this acts as a reminder that this also needs to be changed. I will change it to a warning instead.

tejlmand replied:
> Unless when editing the .config file manually or generating from fragments and some non-Zephyr script?

No, it will always be set, try this for yourself:
```
$ cat build/zephyr/.config |grep LLVM_USE
CONFIG_LLVM_USE_LD=y
# CONFIG_LLVM_USE_LLD is not set
$ sed 's/\(CONFIG_LLVM_USE_LD\)=y/# \1 is not set/g' build/zephyr/.config > tmp; mv tmp build/zephyr/.config
$ cat build/zephyr/.config |grep LLVM_USE
# CONFIG_LLVM_USE_LD is not set
# CONFIG_LLVM_USE_LLD is not set

.... Let's do the build:
$ ninja -Cbuild
ninja: Entering directory `build'
[0/1] Re-running CMake...
Loading Zephyr default modules (Zephyr base (cached)).
....
$ cat build/zephyr/.config |grep LLVM_USE
CONFIG_LLVM_USE_LD=y
# CONFIG_LLVM_USE_LLD is not set
```

As you see, touching `.config` will cause a CMake re-run which triggers a kconfig run that ensures all choices has a value.

> Could someone really add an extra KConfig option without editing this code?

You can extend choices with new entries downstream, for example like it's done here:
https://github.com/nrfconnect/sdk-nrf/blob/5f58fc4ae964e24e5b432bcfcca865f857286f62/Kconfig.nrf#L23-L38
where the Zephyr provided options for an mbedTLS implementation is extended with an extra choice entry in a downstream project.

> Can this FATAL_ERROR really cause invalid .config files to be generated?

No, but a `FATAL_ERROR` in CMake will prevent running `menuconfig`, cause the build file (which contains the `menuconfig` build target) has never been created.

Therefore, mis-configured Kconfig files should not result in fatal errors in CMake cause users will not be able to open `menuconfig` to fix the error.

So in this case, there is no value in throwing an error that a user cannot fix.
Plus it prevents downstream projects to provide extra choice options, and that's making it harder to extend Zephyr on a general basis.

In this particular case, there is probably little risk, but users will copy this pattern else-where, not realizing the damage / limitations it may introduce at other places.

marc-hb commented on cmake/toolchain/llvm/target.cmake:
```diff
@@ -8,3 +8,18 @@ else()
           "Need to enable either CONFIG_LLVM_USE_LD or CONFIG_LLVM_USE_LLD!")
 endif()
+
+# Since kconfigs are not available when generic.cmake
+# is parsed. Setting the target triple for x86 needs to
+# be done here instead.
```
> that's the reason for the split in generic.cmake and target.cmake, so no reason for such comment.

OK but how many other people can easily understand that besides yourself + 1 or 2 others in the entire world? I know it's very hard but please try to put yourself in the shoes of non domain experts. I do NOT have the answer, this is a genuine question, not a rhetorical question. I do not have the answer but if @dcpleung who just spent weeks working on this PR - which makes it much, much more expert than the average - felt the need for this comment then I would tend to trust him that it is useful. Also: how does it hurt? It's only 3 lines and very far from a CMake crash course.

dcpleung replied:
I have removed this but kept the comment on setting `LINKER` in `generic.cmake`, where it mentioned about kconfig not being available during parsing of `generic.cmake`.

tejlmand replied:
> OK but how many other people can easily understand that besides yourself + 1 or 2 others in the entire world?

Then the comment should be the purpose of the generic.cmake / target.cmake files, and not why the triple (in this file) or why setting the linker is deferred to target.cmake (comment in the generic file).

I would be in favor of general comments in top of the file, for example like this:
```diff
+++ b/cmake/toolchain/<toolchain>/generic.cmake
@@ -1,5 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
+# Purpose of the generic.cmake is to define a generic C compiler which can be
+# used for devicetree pre-processing and other pre-processing tasks which must
+# be performed before the target can be determined.
+
```

dcpleung replied:
Added this and removed the other comments.

marc-hb commented on cmake/toolchain/llvm/Kconfig:
```diff
@@ -0,0 +1,19 @@
+	default LLVM_USE_LD
+
+config LLVM_USE_LD
+	bool "LLVM use ld linker"
```
```suggestion
    bool "GNU ld"
```
As already mentioned in this PR `ld` is not a name, it only means generic "linker"

tejlmand replied:
Even better, thanks @marc-hb

dcpleung commented on cmake/toolchain/llvm/generic.cmake:
```diff
@@ -11,13 +11,13 @@ endif()
-endif()
 set(BINTOOLS llvm)
 
+# Note that ${LINKER} is dependent on CONFIG_LLVM_USE_LLD
+# or CONFIG_LLVM_USE_LD, and these kconfigs are not
+# available when this file is parsed. So setting ${LINKER}
+# needs to be deferred to target.cmake.
+
```
I put it here because it deviates from other toolchains' `generic.cmake`, simply trying to avoid confusing future readers.

tejlmand replied:
> I put it here because it deviates from other toolchains' generic.cmake, simply trying to avoid confusing future readers.

I fully understand why you did it, but I believe it would then be better to make a general comment in the top of this file describing the purpose of the file.

tbhI also think we should cleanup the other files, so that generic.cmake and target.cmake contains what they really should contain and not messed like now. But I consider that out-of-scope for this PR.

dcpleung committed:
linker: ld: GNULD_LINKER_IS_BFD to indicate if ld.bfd is used

This adds a new output variable to FindGnuLd.cmake to indicate
if ld.bfd is found. Since we now ask the compilers for their
preferred ld.bfd linker, it may not match using the existing
string equal test to ${CROSS_COMPILE}ld.bfd. So set the new
variable GNULD_LINKER_IS_BFD to true if ld.bfd, and use it to
pass an extra argument to compiler to make it use ld.bfd.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tejlmand commented on cmake/modules/FindGnuLd.cmake:
```diff
@@ -37,8 +42,11 @@ if(NOT EXISTS "${GNULD_LINKER}")
+    set(GNULD_LINKER_IS_BFD TRUE)
+  else()
     find_program(GNULD_LINKER ${CROSS_COMPILE}ld ${LD_SEARCH_PATH})
+    set(GNULD_LINKER_IS_BFD FALSE)
```
is that always true, or just true 99% of the time ?
If we start reporting this, then we should really be sure that it's correct.

In linux, there will almost always an `ld.bfd` and `ld` will usually be a link to the `ld.bfd` exe.
The `almost` is to take into account any obscure distro / home built toolchain where `ld` could be an `ld.bfd` without `ld.bfd` existing (or not in path).

But what about other OS'es and pre-built cross tool-chains ?
Are we sure all conforms to the `<triple>-ld.bfd` or could some be lazy and provide only `<triple>-ld`, for example on Windows where links (the unix way) are not supported.
Such that `<triple>-ld.exe` is in fact a bfd.

I agree that most likely your assumption is correct, but worried about edge-cases.

dcpleung replied:
This is more of a case where we can be certain that the linker is bfd compatible. We do ask the compilers for their preferred `ld.bfd` binary, and if they say there is one, we assume it is bfd compatible. This is simply a case where if we have any doubts, assume the linker is not bfd compatible.

tejlmand replied:
accepted

tejlmand reviewed (commented):
Looks good.

A minor concern to reply to before a +1

dcpleung commented:
Compliance failure is due to the kconfig file being selectively included (based on which toolchain being used), and thus any documentation on those kconfigs generated check errors as clang is not used by default (and thus not the checked).

dcpleung commented:
The last change was simply adding the kconfigs to the compliance check ignore list.

marc-hb commented:
In some conditions, the commit that adds `GNULD_LINKER_IS_BFD` causes very confusing `"-fuse-ld=bfd: unknown flag` failures:
- #56501
